### DISCUSSION TITLE
Is kv_cache being read for anything other than cross_attn with GreedyDecoder?

### DISCUSSION THREAD

**Discussion by mmmkkaaayy** [openai/whisper#1738] [Q&A] [answered]
Based on my debugging, I've seen that `kv_cache` is written to during `self.attn` and `self.cross_attn` calls of the `TextDecoder` forward pass. However, I can only see values from the `kv_cache` being _read_ during `self.cross_attn` (by entering [this](https://github.com/openai/whisper/blob/b38a1f20f4b23f3f3099af2c3e0ca95627276ddf/whisper/model.py#L85-L88) branch, which is not possible for `self.attn` because `xa` is not passed as an arg). 

Otherwise, the only other time I found the `kv_cache` being accessed is [here](https://github.com/openai/whisper/blob/b38a1f20f4b23f3f3099af2c3e0ca95627276ddf/whisper/model.py#L203), where I feel an int counter would be sufficient instead of having to actually save any inferred values. 

Am I missing some other part of the code/conditions under which the `kv_cache` is being read? I see [this](https://github.com/openai/whisper/blob/b38a1f20f4b23f3f3099af2c3e0ca95627276ddf/whisper/decoding.py#L363) line in BeamSearchDecoder, but I'm interested in the GreedyDecoder path.

**Comment by mmmkkaaayy** [comment]
Couple other notes:
* [Appending](https://github.com/openai/whisper/blob/b38a1f20f4b23f3f3099af2c3e0ca95627276ddf/whisper/model.py#L296) to the cache indeed reads the previous value. But is the final result used anywhere? 
* https://github.com/openai/whisper/discussions/489 confirms that the kv_cache is written to in both `self.attn` and `self.cross_attn`, but doesn't directly answer my question about whether the values are being read.

**Reply by alexterminator88** [reply]
.github

**Comment by mmmkkaaayy** [comment, marked as answer]
Oh, I get it now... the return value of `save_to_cache` is actually overriding the return value of the layer it's hooked into, so that's how the values are being read.
